### PR TITLE
feat: add lock escrow

### PR DESCRIPTION
**Pull request by namgold**
Introduced a new lock escrow mechanism supported by the program.

Features:
- `AmmImpl.getLockedAtaAmount` to retrieve locked LP amounts from the ATA mechanism (old mechanism).
- `AmmImpl.getLockedLpAmount` to fetch locked LP amounts from two versions of locking: ATA (old) and escrow (new), and then sum them.
- `AmmImpl.getUserLockEscrow` to obtain the user's lock escrow state.
- `AmmImpl.lockLiquidity` to lock the user's LP into the lock escrow.
- `AmmImpl.claimLockFee` to claim fees from locked LPs in the lock escrow.

### CODE REVIEW

File: ts-client/src/amm/index.ts [hunk 1]
```diff
@@ -1334,6 +1359,145 @@ export default class AmmImpl implements AmmImplementation {
...
+        bVaultLp: this.poolState.bVaultLp,
+        aVaultLpMint: this.vaultA.vaultState.lpMint,
+        bVaultLpMint: this.vaultB.vaultState.lpMint,
+      })
+      .preInstructions(preInstructions)
+      .transaction();
+
+    return new Transaction({
+      feePayer: owner,
+      ...(await this.program.provider.connection.getLatestBlockhash('finalized')),
```
Comment (00xSam) [inline ts-client/src/amm/index.ts:1440]: this should be `confirmed`

File: ts-client/src/amm/index.ts [hunk 2]
```diff
@@ -1334,6 +1359,145 @@ export default class AmmImpl implements AmmImplementation {
...
+        bTokenVault: this.vaultB.vaultState.tokenVault,
+        userAToken: tokenAAta,
+        userBToken: tokenBAta,
+      })
+      .preInstructions(preInstructions)
+      .postInstructions(postInstructions)
+      .transaction();
+    return new Transaction({
+      feePayer: owner,
+      ...(await this.program.provider.connection.getLatestBlockhash('finalized')),
```
Comment (00xSam) [inline ts-client/src/amm/index.ts:1497]: same here

File: ts-client/src/amm/utils.ts [hunk 3]
```diff
@@ -123,6 +124,10 @@ export const getOrCreateATAInstruction = async (
   }
 };
 
+export const deriveLockEscrowPda = (pool: PublicKey, owner: PublicKey, ammProgram: PublicKey) => {
+  return PublicKey.findProgramAddressSync([Buffer.from('lock_escrow'), pool.toBuffer(), owner.toBuffer()], ammProgram);
```
Comment (00xSam) [inline ts-client/src/amm/utils.ts:128]: 'lock_escrow' should put in `SEEDS` in constants.ts

File: ts-client/src/amm/index.ts [hunk 4]
```diff
@@ -1330,10 +1355,149 @@ export default class AmmImpl implements AmmImplementation {
...
+      lockEscrowAccount.unclaimedFeePending,
+      this.poolInfo.virtualPriceRaw,
+    );
+
+    const { tokenAOutAmount, tokenBOutAmount } = this.getWithdrawQuote(unClaimedFee, 0);
+    return {
+      address: lockEscrow,
+      amount: lockEscrowAccount.totalLockedAmount || new BN(0),
+      fee: {
+        tokenAMint: lockEscrowAccount.tokenAMint,
```
Comment (andrewsource147) [inline ts-client/src/amm/index.ts:1378]: we dont have tokenAMint in lockEscrowAccount, should get this from pool account

File: ts-client/src/amm/index.ts [hunk 5]
```diff
@@ -1330,10 +1355,149 @@ export default class AmmImpl implements AmmImplementation {
...
+        tokenBMint: lockEscrowAccount.tokenBMint,
```
Comment (andrewsource147) [inline ts-client/src/amm/index.ts:1379]: Ditto^

**Review by 00xSam — APPROVED** [review]

**Pull request merged**